Add Warning Handling and Dynamic Actions for All Plugins Based on Publish Plugin Process Results

nessid opened the pull request:
## Description

This pull request introduces the `plugins_view_actions` branch, which builds on the features from the `warnings-report_post-process-actions` branch and adds new functionalities.

## Changelog

### Features from `warnings-report_post-process-actions`:
- **Handling of Non-Blocking Publish Errors (Warnings)**:
  - Included non-blocking publish errors (warnings) in the report.
  - Added ability to ignore warnings and proceed with publishing.
  - Introduced the `PublishValidationException` class to handle validation errors and warnings during the publishing process. This class serves as a parent to the pre-existing `PublishValidationError` class and the new `PublishValidationWarning` class.

- **Dynamic Actions in `PublishPluginsProxy`**:
  - Added actions to `PublishPluginsProxy` after plugins have been processed.
  - Enabled handling and filtering of dynamically created actions based on publish plugin process results.
  - Ensured that actions added dynamically during the plugin process are captured and executed appropriately.

- **UI and Visual Enhancements**:
  - Updated color codes and UI elements to enhance visual clarity.

### New Features in `plugins_view_actions`:
- **Plugin Actions in the "Details" Tab**:
  - Added actions to the plugins list in the "Details" tab of the publisher.
  - Enabled actions for all plugins, not just those with errors or warnings.

## Additional Information
- The handling of non-blocking publish errors (warnings) allows users to see warnings in the report and choose to ignore them if needed, proceeding with the publish process. These warnings are managed via `PublishValidationException`.
- The `PublishValidationException` class now serves as the base for specific validation error and warning classes, offering detailed error messages and descriptions using Markdown syntax. This change enhances error handling and user communication in the publishing workflow.
- Actions are now added to `PublishPluginsProxy` after plugins have been processed, allowing for handling and filtering of actions that are created dynamically based on the publish plugin process results and the actions 'on' attribute.
- The new feature in this branch introduces actions in the plugins list within the "Details" tab of the publisher, providing actions for all plugins, including those without errors or warnings.
-  The new features can be tested with the `ayon-studio-addon`. As this package is a under development, you will need to register its plugins manually. Please refer to the README file for instructions.

nessid committed:
feat: warnings report / get actions post plugin process

- report warnings instead of only errors
- get actions by plugin post plugin process to get actions collected during process also
- slight icons and colors modifications

nessid committed:
feat: warnings report / get actions post plugin process

- report warnings instead of only errors
- get actions by plugin post plugin process to get actions collected during process also
- slight icons and colors modifications

nessid committed:
chore: add log levels and warning exception

nessid committed:
chore: color adjustments

remove "info" color and change log warning color to match exception warning

nessid committed:
refactor: PublishValidationReportItem class

add parent class to PublishValidationError and PublishValidationWarning

nessid committed:
chore: adjust colors and icons

adjust colors and icons depending on the plugin proces result

nessid committed:
refactor: extact _determine_icon_and_color from paintEvent

nessid committed:
chore: remove print

nessid committed:
fix: warned plugins

Adapt way to indentify warned plugins to the new way to declare them

nessid committed:
fix: warned plugins

Adapt way to indentify warned plugins to the new way to declare them

nessid committed:
refactor: use original method and variables names

use "error" back instead of "report_item", as now both error and warnings are Exceptions. This allows to stay closer to the original code

nessid committed:
chore: _on_publish_stop

add docstring

nessid committed:
chore: original names

nessid committed:
feat: warned

add warned attribute to LogicItam and InstanceItem to have the right icon on instances in Details tab

nessid committed:
fix: correct if statement

nessid committed:
feat: warned

add ITEM_WARNED_ROLE

nessid committed:
fix: correct errored plugin detection

nessid committed:
doc: correct docstring

nessid committed:
refactor: use original names

use original names from main branch when possible

nessid committed:
revert: original names

Use original names when it is possible

nessid committed:
revert: original names

Use original names when it is possible

nessid committed:
chore: rename variables

nessid committed:
chore: rename class and attributes

replace "report item" with "exception" and "report_type" with "exception_type"

nessid committed:
revert: match with main branch

nessid committed:
fix: add error type indicators to PublishValidationErrorsReport

- Added `has_error_type_errors` and `has_warning_type_errors` parameters to the constructor of `PublishValidationErrorsReport`.
- Updated the docstring to include the new parameters.
- Modified the `from_data` method to deserialize `has_error_type_errors` and `has_warning_type_errors` from the provided data.
- Ensured the `to_data` method includes `has_error_type_errors` and `has_warning_type_errors` in the serialized output.

nessid committed:
revert: restore typos to align with the original codebase

The original codebase's typos were corrected to improve readability. To maintain alignment with the source repository, these changes are being reverted.

nessid committed:
chore: remove unused icon

Removed an icon that was added to the codebase but is no longer in use.

nessid committed:
revert: restore typos to align with the original codebase

The original codebase's typos were corrected to improve readability. To maintain alignment with the source repository, these changes are being reverted.

nessid committed:
revert: restore typos to align with the original codebase

The original codebase's typos were corrected to improve readability. To maintain alignment with the source repository, these changes are being reverted.

nessid committed:
revert: restore typos to align with the original codebase

The original codebase's typos were corrected to improve readability. To maintain alignment with the source repository, these changes are being reverted.

nessid committed:
feat: add plugin_view_actions

wip

nessid committed:
feat: add custom colors to log icons and validation error icons

- Added log level colors for debug, info, warning, error, and critical log levels in the stylesheet.
- Implemented color change for validation error and warning icons in the report viewer.
- Updated `LogIconFrame` and `ValidateErrorIconFrame` to use added log-level colors.
- Improved the visibility management of log icons and validation error icons.
- Updated `settings.json` to include new color definitions for log levels and validation states.

nessid committed:
chore: remove unused import

nessid committed:
feat: wip add actions

nessid committed:
feat: draw plugins actions icons

nessid committed:
feat: add action icon

use icon from pyblish-lite

nessid committed:
refactor: simplify attribute default assignment to ensure list

Use `getattr(plugin, "actions", [])` instead of `getattr(plugin, "actions", None) or []` to always ensure `actions` is a list.

nessid committed:
refactor: optimize action handling and tooltip setting in item_paint

Refactor the handling of action icons by introducing `active_actions` to filter relevant actions more efficiently. Ensure consistent initialization of `plugin_actions` to an empty list if not present. Streamline the tooltip setting logic for elided labels.

nessid committed:
feat: consider action.active in active_actions filtering

Include action.active attribute in the conditions for filtering active actions in PluginDelegate.

nessid committed:
refactor: filter active actions in PluginItem

Add filtering logic to PluginItem to determine active actions based on the plugin's error and warning states.

nessid committed:
refactor: filter active actions moved in PluginItem

 Filtering logic moved to PluginItem to determine active actions based on the plugin's error and warning states.

nessid committed:
refactor: filter plugins actions

nessid committed:
refactor: filter plugins actions

nessid committed:
refactor: add warned and errored to plugin_data_item in PublishReportMaker

nessid committed:
refactor: move warned and errored to plugin_data_item in PublishReportMaker

nessid committed:
refactor(publisher): reorganize action methods and update plugin data structure

- Moved action methods from `PublishReportMaker` to `PublishPluginProxy`.
- Added `warned`, `errored`, and `action_items` to `plugin_data` in `PublishReportMaker` with methods to set them: `set_plugin_warned`, `set_plugin_errored`, `set_plugin_action_items`
- Renamed the key "actions" to "action_items" in the dictionary within the `PublishReportMaker` class and related classes.
- Updated the `PluginItem` class in `report_items.py` to use `action_items` instead of `actions`.
- Adapted the `_process_and_continue`, `report_page`, and `report_items` methods to accommodate these changes.

nessid committed:
refactor: add action buttons in plugins view and support controller to run actions

support controller in GroupItemDelegate, PublisherWindow and PublishReportViewerWidget to run actions on ActionButton clicked

nessid committed:
refactor: remove obsolete filter mapping and applicable_filters logic

The code related to filter_mapping and applicable_filters has been removed.

This cleanup was done because the actions in PublishPluginsProxy are now already filtered, rendering this code unnecessary.

nessid committed:
fix: correct default value for is_validation_error in LogItem

Changed the default value for `is_validation_error` in `LogItem` class to `False`. This ensures that a log item is not erroneously marked as a validation error if the `is_validation_error` key is absent in `log_item_data`.
This fix provides a more accurate default behavior, reducing the chance of misclassifying log items.

nessid committed:
fix: correct default value and simplify LogItem initialization

Changed the default value for `is_validation_error` in `LogItem` class to `False`. This ensures that a log item is not erroneously marked as a validation error if the `is_validation_error` key is absent in `log_item_data`.

Simplified the LogItem class initialization by directly using the `get` method for `is_validation_error` and `is_validation_warning`.

nessid committed:
chore: fix typo

nessid committed:
refactor: align branch with original code by removing non-essential changes

KernAttila commented on client/ayon_core/tools/publisher/publish_report_viewer/delegates.py:
```diff
@@ -118,8 +127,8 @@ def _get_minus_pixmap(cls, size):
-            return QtGui.QColor(colors["error"])
+        if name == "error" or name == "warning":
+            return QtGui.QColor(colors[name])
         return QtGui.QColor(QtCore.Qt.white)
```
```suggestion
     return QtGui.QColor(colors.get(name, QtCore.Qt.white))
```

nessid replied:
nice. It's a bit further from the original code, but it's better. They should do that too :)

nessid committed:
refactor(error-handling): rename warning type errors to non-blocking errors

Renamed warning type errors to non-blocking errors for better clarity.
Blocking errors are now clearly differentiated from non-blocking errors.
This change addresses confusion caused by treating errors as warnings.

nessid committed:
refactor: align branch with original code by removing non-essential changes

nessid committed:
refactor(error-handling): rename warning type errors to non-blocking errors

Renamed warning type errors to non-blocking errors for better clarity.
Blocking errors are now clearly differentiated from non-blocking errors.
This change addresses confusion caused by treating errors as warnings.

nessid committed:
refactor: remove "warning" state from _set_success_property as validation warnings no longer exist

Removed the "warning" state from the initial state of _set_success_property, which controls the appearance of the publish_frame. This change was made because validation warnings have been eliminated, leaving only blocking and non-blocking errors.

nessid committed:
chore: reduce discrepancies with main branch

nessid committed:
refactor: simplify error handling logic in state management

Progress visibility is set to False once when errors are detected, regardless of whether errors are blocking or non-blocking.

nessid committed:
feature: add `Validation non-blocking error`state for publish_frame

nessid committed:
feat: use checked icon for all record logs

EnisNewborn commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -2509,6 +2657,7 @@ def _process_and_continue(self, plugin, instance):
             result["is_validation_error"] = has_validation_error
 
         self._publish_report.add_result(result)
+        self._publish_report.set_plugin_action_items(self._publish_plugins_proxy.get_plugin_action_items(plugin.id))
```
number of characters per lines ;)

nessid replied:
my mistake—the default character limit in PyCharm was set to 120 characters. I will need to review my entire code I guess...

EnisNewborn commented on client/ayon_core/tools/publisher/publish_report_viewer/delegates.py:
```diff
@@ -118,8 +127,8 @@ def _get_minus_pixmap(cls, size):
     def _get_icon_color(cls, name):
-        if name == "error":
-            return QtGui.QColor(colors["error"])
+        if name == "error" or name == "warning":
```
`if name in ["error", "warning"]:`

nessid replied:
done

EnisNewborn commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -2250,12 +2371,28 @@ def validate(self):
+        else:
+            self.log.info("Resuming publishing after ignoring non-blocking errors.")
+            if self._btn_clicked == "validate":
+                self.validate()  # or self.publish() depending on context
```
no end line comment

nessid replied:
done

EnisNewborn commented on client/ayon_core/tools/publisher/publish_report_viewer/delegates.py:
```diff
@@ -213,16 +227,31 @@ def item_paint(self, painter, option, index):
+        if index.data(PLUGIN_ACTIONS_ROLE):
+            action_rect.setWidth(action_rect.height() + 5)
+            action_rect.translate(
+                bg_rect.width() - action_rect.width(),  # Adjusted position to include margin
```
no end line comment

nessid replied:
done

EnisNewborn commented on client/ayon_core/tools/publisher/publish_report_viewer/delegates.py:
```diff
@@ -213,16 +227,31 @@ def item_paint(self, painter, option, index):
+                bg_rect.width() - action_rect.width(),  # Adjusted position to include margin
+                0
+            )
+            self.action_icon_rects[index] = action_rect  # Save the action icon rect for mouse click detection
```
no end line comment

nessid replied:
done

EnisNewborn commented on client/ayon_core/tools/publisher/publish_report_viewer/delegates.py:
```diff
@@ -247,7 +280,25 @@ def item_paint(self, painter, option, index):
+        # Draw action icon
+        if index.data(PLUGIN_ACTIONS_ROLE):
+            action_icon = get_pixmap("adn")
+            icon_size = action_rect.height() - 2  # Size of the action icon
```
no end line comment

nessid replied:
done

EnisNewborn commented on client/ayon_core/tools/publisher/publish_report_viewer/delegates.py:
```diff
@@ -247,7 +280,25 @@ def item_paint(self, painter, option, index):
+                                             QtCore.Qt.SmoothTransformation)
+
+            # Adjust label_rect to avoid overlap with action icon
+            label_rect.setWidth(label_rect.width() - action_rect.width() - 10)  # Added margin
```
no end line comment

nessid replied:
done

EnisNewborn commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -655,7 +716,15 @@ def from_data(cls, data):
         ]
-        return cls(error_items, plugin_action_items)
+
+        has_blocking_errors = data["has_blocking_errors"]
```
why not using .get() here ? (real question)

nessid replied:
I used direct dictionary access to maintain consistency with the pre-existing code. Since I write my data simultaneously with the pre-existing data, missing keys would indicate an issue with my added code.

EnisNewborn reviewed (changes requested):
Few little updates.
Kern will have a look too

nessid committed:
refactor: simplify conditional check and clean up code

Simplified the conditional check for the 'name' variable by using a list instead of multiple 'or' conditions. Removed end-of-line comments and ensured lines do not exceed 79 characters.

nessid committed:
style: shorten lines to 79 characters
docs: add docstrings for better documentation

Shortened the length of lines to 79 characters to improve code readability and adhere to PEP 8 guidelines.
Added docstrings to functions and classes for better documentation and understanding of the code.

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -340,27 +377,39 @@ class PublishPluginsProxy:
             plugin_id = plugin.id
-            plugins_by_id[plugin_id] = plugin
+            self._plugins_by_id[plugin_id] = plugin
+
```
Shouldn't we call self.set_plugin_actions() here ?
In the original code, the variables are set during init, 
could you explain if there is potential side-effects by not doing it at init ?

nessid replied:
I'm calling it after the plugins have been processed since it needs the results to retrieve the actions added during processing and any validation errors that play a role in filtering the actions.

KernAttila replied:
Okay, sounds good !

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -711,23 +850,32 @@ def add_error(self, plugin, error, instance):
-            else:
-                plugin_label = plugin.__name__
-            error.title = plugin_label
+        self._ensure_validation_title(plugin, error)
```
picky:
renaming this to `_validate_error_title` seems less confusing.

nessid replied:
to be even more clear about what it's doing, could we call it
`_set_error_title_if_missing`?

nessid replied:
I noticed that – if I'm not mistaken – I could restore PublishValidationErrors to its initial state. Since no modifications were still needed, removing this method, which didn't exist originally, restored the class completely.

KernAttila replied:
I'm okay with that :)

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -664,6 +784,8 @@ class PublishValidationErrors:
     def __init__(self):
         self._plugins_proxy = None
         self._error_items = []
+        self._blocking_errors = []
+        self._non_blocking_errors = []
```
I don't find any particular use of this variable except for the property.
In my opinion, we pass too much data around between objects.
What interests us here are the two properties below, `has_blocking_errors` and `has_non_blocking_errors`.

KernAttila replied:
I get that you wanted to match the current behavior for the boolean tests, but we can avoid it (see comments below)

KernAttila replied:
If there is no other use for these two lists, you can get rid of them.

nessid replied:
That's very true. For some reason, I thought I couldn't get rid of those lists. Now I see I obviously can. 🤷‍♀️

nessid replied:
I removed self.has_blocking_errors and self.has_non_blocking_errors from the PublishValidationErrors class and added this to PublishValidationErrorsReport:

```
def __init__(self, error_items, plugin_action_items):
    self._error_items = error_items
    self._plugin_action_items = plugin_action_items
    self.has_blocking_errors = False
    self.has_non_blocking_errors = False

    for error in self._error_items:
        if error.is_blocking:
            self.has_blocking_errors = True
        else:
            self.has_non_blocking_errors = True

        if self.has_blocking_errors and self.has_non_blocking_errors:
            break
```
is it ok?

KernAttila replied:
That's perfect :)

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -675,6 +797,18 @@ def has_errors(self):
+    def has_blocking_errors(self):
+        """At least one blocking error was added."""
+
+        return bool(self._blocking_errors)
```
This is a bit heavier to compute, but avoids having to pass that much data for this single purpose.
```suggestion
        return any(error.is_blocking for error in self._error_items)
```

nessid replied:
I used this in `PublishValidationErrorsReport` instead of `PublishValidationErrors`. This change was possible because the data was only meant to be passed to `PublishValidationErrorsReport` and wasn't necessary in `PublishValidationErrors`, which has been reverted to its original state from the upstream repository)

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -675,6 +797,18 @@ def has_errors(self):
+    def has_non_blocking_errors(self):
+        """At least one non-blocking error was added."""
+
+        return bool(self._non_blocking_errors)
```
Same here :
```suggestion
        return any(not error.is_blocking for error in self._error_items)
```

nessid replied:
Same, I used this in PublishValidationErrorsReport and removed it from PublishValidationErrors

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -1453,6 +1613,9 @@ def _set_publish_has_crashed(self, value):
     def _get_publish_has_validation_errors(self):
         return self._publish_has_validation_errors
 
+    def _get_publish_has_validation_blocking_errors(self):
+        return self._publish_has_validation_blocking_errors
+
```
I might be wrong, but I feel this should totally replace `_get_publish_has_validation_errors`.
In `stop_publish()`, we are not yet using blocking errors, but all errors, blocking or not.
What happens if you have only non-blocking errors during publish ? As they will appear in the error list, thus blocking the publish.

This feature is a change of behavior, not an additional one.
Before we were stuck with all errors. Now we stop on blocking errors but continue on non-blocking ones.

It makes sense to stop at any error type with the UI, but in batch mode, it must fail only if there is blocking errors, bypassing non-blocking ones.

nessid replied:
The `_is_interactive` method has been added to determine if the code is running in interactive or batch mode, in order to adapt the behavior and to stop the publish process for non-blocking errors in interactive mode only.
It still needs to be tested in batch.

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -1632,6 +1807,9 @@ def __init__(self, headless=False):
         self._hierarchy_model = HierarchyModel(self)
 
+        self._ignore_non_blocking_errors = False
+        self._btn_clicked = None
```
Thank you ^^

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -2472,8 +2670,15 @@ def _publish_iterator(self):
 
-    def _add_validation_error(self, result):
+    def _add_validation_error(self, result: dict) -> None:
+        if not result["is_blocking"] and self._ignore_non_blocking_errors:
+            return
```
Add a log here:
```suggestion
        if not result["is_blocking"] and self._ignore_non_blocking_errors:
            self.log.info(f"Ignoring non-blocking error {result['plugin']}")
            return
```

nessid replied:
I recycled here the _`set_error_title_if_missing` method that I removed from `PublishValidationErrors` to use it for the log:
```
    @staticmethod
    def _set_error_title_if_missing(plugin, error):
        """Ensure that the validation report item has a title."""
        if not error.title:
            error.title = getattr(plugin, 'label',
                                  plugin.__name__)

    def _add_validation_error(self, result: dict) -> None:
        self._set_error_title_if_missing(result["plugin"], result["error"])
        if not result["is_blocking"] and self._ignore_non_blocking_errors:
            self.log.info(
                f"Ignoring non-blocking error: {result['error'].title}")
            return
```
The verification that is still made in the original code in `PublishValidationErrors` will be useless but in the log we'll have the same error title that we have in the Report page of the Publisher.

KernAttila commented on client/ayon_core/tools/publisher/control_qt.py:
```diff
@@ -174,6 +174,10 @@ def remote_events_handler(self, event_data):
             self.publish_has_validation_errors = event["value"]
             return
```
Same here, it seems to become obsolete with the new behavior.

KernAttila commented on client/ayon_core/tools/publisher/publish_report_viewer/delegates.py:
```diff
@@ -118,8 +127,8 @@ def _get_minus_pixmap(cls, size):
     def _get_icon_color(cls, name):
-        if name == "error":
-            return QtGui.QColor(colors["error"])
+        if name in ["error", "warning"]:
+            return QtGui.QColor(colors[name])
         return QtGui.QColor(QtCore.Qt.white)
```
You can be a bit more abstract here.
```suggestion
        return QtGui.QColor(colors.get(name, QtCore.Qt.white))
```

nessid replied:
done

KernAttila commented on client/ayon_core/tools/publisher/publish_report_viewer/report_items.py:
```diff
@@ -12,17 +12,9 @@ def __init__(self, plugin_data):
-
-        self.errored = errored
+        self.actions = plugin_data["action_items"]
+        self.errored = plugin_data["errored"]
```
Do we want to get rid of the "instances_data" loop here ?

nessid replied:
I think so, but I might be wrong... The information is already available without the loop, and since I was already adding "action_items" and "is_blocking" to the plugin data item, it made sense to me to pass "errored" at the same time instead of retrieving it separately. I don't remember if that was the only reason, but looking at the code, I looks like it.

nessid replied:
can I resolve this one?

KernAttila reviewed (changes requested):
Overall, I like the implementation.
I have some comments on redundancy.
I didn't spend too much time on the UI code, it seems good :)

nessid committed:
style: format non-original code according to PEP8

Preserve original code as this is a fork, allowing for easier future updates from the upstream repository.

nessid committed:
refactor(validation-errors): simplify and optimize error handling

Refactored the `PublishValidationErrorsReport` and `PublishValidationErrors` classes to simplify the error handling and initialization processes. The changes include:

- Introduced logic in the constructor of `PublishValidationErrorsReport` to set `has_blocking_errors` and `has_non_blocking_errors` based on the provided error items.
- Removed unnecessary lists and attributes to streamline the classes.

`PublishValidationErrors` class is restored to its original state to simplify future synchronizations with the main repository.

- Implement the assignment of a title for errors that don't have one earlier in the process, enabling logging of information with the title of the ignored errors.

nessid committed:
refactor(icon): abstract _get_icon_color method using dict.get()

Refactor the _get_icon_color method by replacing conditional statements with a single call to dict.get().

nessid committed:
feat: add interactive mode detection and update publish logic

- Add `_is_interactive` method to determine if the script is running in interactive mode or batch mode.
- Update publish and validation logic to handle interactive and batch modes differently:
  - In interactive mode, stop publishing for any validation errors.
  - In batch mode, stop publishing only for blocking validation errors.

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -415,6 +464,52 @@ def _create_action_item(self, action, plugin_id):
+                     )
+            )
+
+        return [action for action in actions if is_action_active(action)]
```
for info: another syntax for this:

```suggestion
        return list(filter(is_action_active, actions))
```

KernAttila replied:
No impact on performance, it is just a matter of taste. Do as you please, it was for info only :)

nessid replied:
it's concise and clear enough. I like it :)

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -2389,17 +2549,57 @@ def run_action(self, plugin_id, action_id):
+        elif host == 'houdini':
+            import hou
+            return hou.isUIAvailable()
+        else:
+            raise ValueError(f"Unknown host environment: {host}")
```
Let's add Blender for good measure :)
```suggestion
        elif host == 'blender':
            import bpy
            return not bpy.app.background
        else:
            raise ValueError(f"Unknown host environment: {host}")
```

nessid replied:
added 👍

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -2389,17 +2549,57 @@ def run_action(self, plugin_id, action_id):
+            # errors if validation has passed
+            stop_publish_conditions = (
+                    self.publish_has_validated
+                    and self.publish_has_validation_errors
```
Shouldn't we add blocking errors also ?
```suggestion
                    and (self.publish_has_validation_errors or self.publish_has_validation_blocking_errors)
```

KernAttila replied:
and maybe we need to use `_ignore_non_blocking_errors` here also ?

KernAttila replied:
Ignore this, Enis and I looked at it and it appears the publish_has_validation_blocking_errors is a subset of publish_has_validation_errors.
publish_has_validation_errors contains blocking and non-blocking.

KernAttila commented on client/ayon_core/tools/publisher/control.py:
```diff
@@ -2472,12 +2672,20 @@ def _publish_iterator(self):
+                if self._is_interactive:
+                    # In interactive mode, stop publish for
+                    # any validation errors
+                    if self.publish_has_validation_errors:
```
Same here, we might want to use non blocking errors and _ignore_non_blocking_errors

KernAttila replied:
Same, ignore this ^^

KernAttila reviewed (changes requested):
Just a minor addition for blender :)
Thanks for all the changes and this new feature !

nessid committed:
feat: add blender interactive check

KernAttila reviewed (approved):
All good for me, you can merge ;)
